wallet/wallet.go

```diff
@@ -838,6 +838,148 @@ func (w *Wallet) syncToBirthday() (*waddrmgr.BlockStamp, error) {
+
+	// We'll initialize the recovery manager with a default batch size of
+	// 2000.
+	recoveryMgr := NewRecoveryManager(
+		w.recoveryWindow, recoveryBatchSize, w.chainParams,
+	)
+
+	// In the event that this recovery is being resumed, we will need to
+	// repopulate all found addresses from the database. For basic recovery,
+	// we will only do so for the default scopes.
+	scopedMgrs, err := w.defaultScopeManagers()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	txmgrNs := tx.ReadBucket(wtxmgrNamespaceKey)
+	credits, err := w.TxStore.UnspentOutputs(txmgrNs)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	if err := recoveryMgr.Resurrect(ns, scopedMgrs, credits); err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	// Now, we can proceed with adding blocks to the recovery manager's
+	// batch. We'll make sure to add all blocks until tip to ensure we
+	// recover all of our funds.
+	var latestCheckptHeight int32
+	if len(w.chainParams.Checkpoints) > 0 {
+		latestCheckptHeight = w.chainParams.
+			Checkpoints[len(w.chainParams.Checkpoints)-1].Height
+	}
+	_, bestHeight, err := chainClient.GetBestBlock()
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	// isCurrent is a helper function that we'll use to determine if the
+	// chain backend is currently synced.
+	isCurrent := func(bestHeight int32) bool {
+		switch c := chainClient.(type) {
+		case *chain.NeutrinoClient:
+			return c.CS.IsCurrent()
+		}
+		return bestHeight >= latestCheckptHeight
+	}
+
+	for height := birthdayStamp.Height; height <= bestHeight; height++ {
+		// If we've reached our best height, wait until we're current to
+		// ensure we reach all blocks until tip.
+		for height == bestHeight && !isCurrent(bestHeight) {
+			time.Sleep(100 * time.Millisecond)
+			_, bestHeight, err = chainClient.GetBestBlock()
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+		}
+
+		// Retrieve the hash and timestamp of the current block and add
+		// it to the batch.
+		hash, err := chainClient.GetBlockHash(int64(height))
+		if err != nil {
+			tx.Rollback()
+			return err
+		}
+		header, err := chainClient.GetBlockHeader(hash)
+		if err != nil {
+			tx.Rollback()
+			return err
+		}
+
+		recoveryMgr.AddToBlockBatch(hash, height, header.Timestamp)
+
+		// We'll checkpoint our current batch every 2K blocks, so we'll
+		// need to start a new database transaction. If our current
+		// batch is empty, then this will act as a NOP.
+		if height%recoveryBatchSize == 0 {
+			err := w.recoverDefaultScopes(
+				chainClient, tx, ns, recoveryMgr.BlockBatch(),
+				recoveryMgr.State(),
+			)
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+
+			if err := tx.Commit(); err != nil {
+				tx.Rollback()
+				return err
+			}
+
+			tx, err = w.db.BeginReadWriteTx()
+			if err != nil {
+				return err
+			}
+			ns = tx.ReadWriteBucket(waddrmgrNamespaceKey)
+		}
+	}
+
+	// Now that we've reached the chain tip, we can process our final batch
```

Comment:
Isn't the assumption that we first do the birthday sync, _then_ recover if needed, and finally sync to the regular chain? The downside of this split up, is that we'll now scan the chain entirely twice, vs doing it once. Additionally, each time we restart the recovery logic will need to start from scratch, rather than pick up where it left off as it does now.